core/src/impl/Kokkos_Atomic_Exchange.hpp

```diff
@@ -419,6 +420,89 @@ void atomic_assign( volatile T * const dest_v , const T val )
 #endif
 #endif
 
+// clang-format on
+#if defined( KOKKOS_ENABLE_HIP )
+__inline__ __device__
+int atomic_exchange( volatile int * const dest , const int val )
+{
+  return atomicExch(const_cast<int *>(dest), val);
+}
+
+__inline__ __device__ unsigned int
+atomic_exchange(volatile unsigned int *const dest, const unsigned int val)
+{
+  return atomicExch(const_cast<unsigned int *>(dest), val);
+}
+
+__inline__ __device__ unsigned long long int
+atomic_exchange(volatile unsigned long long int *const dest,
+                const unsigned long long int val)
+{
+  return atomicExch(static_cast<unsigned long long *>(dest), val);
+}
+
+/** \brief  Atomic exchange for any type with compatible size */
+template <typename T>
+__inline__ __device__ T atomic_exchange(
+    volatile T *const dest,
+    typename std::enable_if<sizeof(T) == sizeof(int), const T &>::type
+        val)
+{
+  int tmp = atomicExch(static_cast<int *>(const_cast<T *>(dest)),
+                       *static_cast<int *>(const_cast<T *>(&val)));
+  return static_cast<T &>(tmp);
+}
+
+template <typename T>
+__inline__ __device__ T atomic_exchange(
+    volatile T *const dest,
+    typename std::enable_if<
+        sizeof(T) != sizeof(int) && sizeof(T) == sizeof(unsigned long long int),
+        const T &>::type val)
+{
+  typedef unsigned long long int type;
+
+  type tmp = atomicExch(static_cast<type *>(const_cast<T *>(dest)),
+                        *static_cast<type *>(const_cast<T *>(&val)));
+  return static_cast<T &>(tmp);
+}
+
+/** \brief  Atomic exchange for any type with compatible size */
+template <typename T>
+__inline__ __device__ void
+atomic_assign(volatile T *const dest,
+              typename Kokkos::Impl::enable_if<sizeof(T) == sizeof(int),
+                                               const T &>::type val)
+{
+  atomicExch(static_cast<int *>(const_cast<T *>(dest)),
+             *static_cast<int *>(const_cast<T *>(&val)));
```

Comment:
If new encounter more of these in the future we should introduce function helpers to perform these cast